### PR TITLE
[SPARK-35561][SQL] Remove leading zeros from empty static number type partition

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?

This PR removes leading zeros from static number type partition when we insert into a partition table with empty partitions.

create table 

    CREATE TABLE `table_int` ( `id` INT, `c_string` STRING, `p_int` int)
    USING parquet PARTITIONED BY (p_int);

insert 

    insert overwrite table table_int partition (p_int='00011')
    select 1, 'c string'
    where true ;
    
|partition|
|---------|
|p_int=11|

    insert overwrite table table_int partition (p_int='00012')
    select 1, 'c string'
    where false ;

|partition|
|---------|
|p_int=00012|

### Why are the changes needed?

This PR creates consistent result when insert empty or non-empty partition

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Add Unit test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 1]
```diff
@@ -351,10 +351,24 @@ object PartitioningUtils {
    */
   def getPathFragment(spec: TablePartitionSpec, partitionSchema: StructType): String = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:352]: Can we use the same code path with the one when it's non-empty?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:352]: cc @MaxGekk FYI

Comment (dgd-contributor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:352]: > Can we use the same code path with the one when it's non-empty?

I think no. The below code separate code path for update partition metadata of updated partition (non-empty) and empty

https://github.com/apache/spark/blob/badb0393d46d7aef90710e51e233fb5077977423/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L189-L198

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:352]: Can you factor out PartitioningUtils.parsePartitions and share the same casting logic?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 2]
```diff
@@ -351,10 +351,24 @@ object PartitioningUtils {
...
-      escapePathName(field.name) + "=" + getPartitionValueString(spec(field.name))
+      escapePathName(field.name) + "=" +
+        getPartitionValueString(
+          removeLeadingZerosFromNumberTypePartition(spec(field.name), field.dataType))
     }.mkString("/")
   }
 
+  def removeLeadingZerosFromNumberTypePartition(value: String, dataType: DataType): String = {
+    dataType match {
+      case ByteType => value.toByte.toString
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:362]: Would it not be sufficient to call .toLong.toString for any integer type? might be simpler.
Also, would float/double ever be used as a partition key?

Comment (dgd-contributor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:362]: > Would it not be sufficient to call .toLong.toString for any integer type? might be simpler.

Yes, i will modify it.
> Also, would float/double ever be used as a partition key?

Fload/Double type is rarely used as a partition key, but i think we should handle them

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala (hunk 1 above)
Comment (dgd-contributor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:352]: PartitioningUtils.parsePartitions use castPartValueToDesiredType to cast multiple type of partition. I think in this issue, we only need cast number type (or remove leading zeros), the other types will be handled by org.apache.spark.sql.catalyst.expressions.CastBase

**Comment by srowen** [comment]
Jenkins test this please

**Comment by dgd-contributor** [comment]
@srowen @HyukjinKwon Do you think this change is ready to merge?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 3]
```diff
@@ -351,10 +351,20 @@ object PartitioningUtils {
...
   def getPathFragment(spec: TablePartitionSpec, partitionSchema: StructType): String = {
     partitionSchema.map { field =>
-      escapePathName(field.name) + "=" + getPartitionValueString(spec(field.name))
+      escapePathName(field.name) + "=" +
+        getPartitionValueString(
+          removeLeadingZerosFromNumberTypePartition(spec(field.name), field.dataType))
     }.mkString("/")
   }
 
+  def removeLeadingZerosFromNumberTypePartition(value: String, dataType: DataType): String = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:360]: I didn't take a close look but I do think we should deduplicate these codes. At the very least you can reuse cast expression

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:360]: This comment is still outstanding - any possibility to refactor?

Comment (dgd-contributor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: I tried to refactor and add an unit test. Could you please review it again? Thank you.

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Merged to master

**Comment by cloud-fan** [comment]
Do we have an end-to-end test? It's not clear to me how to trigger this bug. If a partition is empty, table scan returns nothing, and people can't see the partition values.

**Comment by cloud-fan** [comment]
nvm, I saw it from the JIRA ticket. `show partitions` can trigger this bug.

**Pull request closed without merging**